File: OpenBurnBarCore/Sources/OpenBurnBarMedia/MediaPacketCodec.swift
import Foundation

/// Length-prefixed binary codec for Mercury media frames. Same outer layout
/// as `IrohRelayFrameCodec` (4-byte big-endian length prefix) so the two
/// codecs can share the underlying QUIC stream framing and an audit reader
/// can locate frame boundaries without knowing whether the payload is JSON
/// chat or a binary media frame.
///
/// ```
/// [ u32 BE total payload length ][ MediaFrame.headerByteCount header ][ payload ]
/// ```
public struct MediaPacketCodec: Sendable {
    public enum CodecError: Error, Equatable, Sendable {
        case envelopeTooShort
        case payloadTooLarge(actual: Int, max: Int)
        case headerTruncated
        case unknownKind(UInt8)
        case cursorTruncated
    }

    /// Hard ceiling on a single media frame. Matches the iroh-blobs default
    /// chunk size so a packet that exceeds it is almost certainly a producer
    /// bug or a hostile peer.
    public static let defaultMaxPayloadBytes: Int = 256 * 1024

    public let maxPayloadBytes: Int

    public init(maxPayloadBytes: Int = MediaPacketCodec.defaultMaxPayloadBytes) {
        self.maxPayloadBytes = maxPayloadBytes
    }

    public func encode(_ frame: MediaFrame) throws -> Data {
        let cursorBytes: Int
        if frame.flags.contains(.hasCursorMetadata) && frame.cursor != nil {
            cursorBytes = MediaFrame.cursorMetadataByteCount
        } else {
            cursorBytes = 0
        }
        let totalPayloadCount = MediaFrame.headerByteCount + cursorBytes + frame.payload.count
        guard totalPayloadCount <= maxPayloadBytes else {
            throw CodecError.payloadTooLarge(actual: totalPayloadCount, max: maxPayloadBytes)
        }

        var envelope = Data(capacity: 4 + totalPayloadCount)
        appendUInt32BigEndian(UInt32(totalPayloadCount), to: &envelope)

        // Header
        envelope.append(frame.kind.rawValue)
        envelope.append(frame.flags.rawValue)
        appendUInt32BigEndian(frame.gopID, to: &envelope)
        appendUInt32BigEndian(frame.frameIndex, to: &envelope)
        appendUInt64BigEndian(frame.presentationTimestampMillis, to: &envelope)

        // Optional cursor extension. The producer may set `hasCursorMetadata`
        // without populating `cursor`; in that case we honor the flag the
        // producer set and write 0,0 rather than silently dropping the
        // extension — so receivers can round-trip frames they cannot yet
        // interpret without losing the bit.
        if frame.flags.contains(.hasCursorMetadata) {
            let cursor = frame.cursor ?? MediaFrame.CursorMetadata(x: 0, y: 0)
            appendInt16BigEndian(cursor.x, to: &envelope)
            appendInt16BigEndian(cursor.y, to: &envelope)
        }

        envelope.append(frame.payload)

        return envelope
    }

    public func decode(_ envelope: Data) throws -> (frame: MediaFrame, consumed: Int) {
        let lengthPrefixBytes = 4
        guard envelope.count >= lengthPrefixBytes + MediaFrame.headerByteCount else {
            throw CodecError.envelopeTooShort
        }

        // Re-base the input view so byte indices start at zero. Foundation's
        // `Data.removeFirst(_:)` does not normalize indices and the same
        // bug bit `IrohRelayFrameCodec` historically — copying through a
        // raw-buffer view is the canonical fix in this repo.
        let normalized: Data = envelope.withUnsafeBytes { buffer in
            Data(buffer)
        }

        let totalPayloadCount = Int(readUInt32BigEndian(from: normalized, at: 0))
        guard totalPayloadCount <= maxPayloadBytes else {
            throw CodecError.payloadTooLarge(actual: totalPayloadCount, max: maxPayloadBytes)
        }
        let totalEnvelopeBytes = lengthPrefixBytes + totalPayloadCount
        guard normalized.count >= totalEnvelopeBytes else {
            throw CodecError.headerTruncated
        }

        let headerStart = lengthPrefixBytes
        let kindByte = normalized[headerStart]
        guard let kind = MediaFrame.Kind(rawValue: kindByte) else {
            throw CodecError.unknownKind(kindByte)
        }
        let flags = MediaFrame.Flags(rawValue: normalized[headerStart + 1])
        let gopID = readUInt32BigEndian(from: normalized, at: headerStart + 2)
        let frameIndex = readUInt32BigEndian(from: normalized, at: headerStart + 6)
        let pts = readUInt64BigEndian(from: normalized, at: headerStart + 10)

        var afterHeader = headerStart + MediaFrame.headerByteCount
        var cursor: MediaFrame.CursorMetadata? = nil
        if flags.contains(.hasCursorMetadata) {
            let cursorEnd = afterHeader + MediaFrame.cursorMetadataByteCount
            guard cursorEnd <= lengthPrefixBytes + totalPayloadCount else {
                throw CodecError.cursorTruncated
            }
            let cursorX = readInt16BigEndian(from: normalized, at: afterHeader)
            let cursorY = readInt16BigEndian(from: normalized, at: afterHeader + 2)
            cursor = MediaFrame.CursorMetadata(x: cursorX, y: cursorY)
            afterHeader = cursorEnd
        }

        let payloadStart = afterHeader
        let payloadEnd = lengthPrefixBytes + totalPayloadCount
        let payload = normalized.subdata(in: payloadStart..<payloadEnd)

        let frame = MediaFrame(
            kind: kind,
            flags: flags,
            gopID: gopID,
            frameIndex: frameIndex,
            presentationTimestampMillis: pts,
            cursor: cursor,
            payload: payload
        )
        return (frame, totalEnvelopeBytes)
    }

    // MARK: Wire helpers

    private func appendUInt32BigEndian(_ value: UInt32, to data: inout Data) {
        var be = value.bigEndian
        withUnsafeBytes(of: &be) { data.append(contentsOf: $0) }
    }

    private func appendUInt64BigEndian(_ value: UInt64, to data: inout Data) {
        var be = value.bigEndian
        withUnsafeBytes(of: &be) { data.append(contentsOf: $0) }
    }

    private func appendInt16BigEndian(_ value: Int16, to data: inout Data) {
        var be = value.bigEndian
        withUnsafeBytes(of: &be) { data.append(contentsOf: $0) }
    }

    private func readUInt32BigEndian(from data: Data, at offset: Int) -> UInt32 {
        var raw: UInt32 = 0
        _ = withUnsafeMutableBytes(of: &raw) { dest in
            data.copyBytes(to: dest, from: offset..<(offset + 4))
        }
        return UInt32(bigEndian: raw)
    }

    private func readUInt64BigEndian(from data: Data, at offset: Int) -> UInt64 {
        var raw: UInt64 = 0
        _ = withUnsafeMutableBytes(of: &raw) { dest in
            data.copyBytes(to: dest, from: offset..<(offset + 8))
        }
        return UInt64(bigEndian: raw)
    }

    private func readInt16BigEndian(from data: Data, at offset: Int) -> Int16 {
        var raw: Int16 = 0
        _ = withUnsafeMutableBytes(of: &raw) { dest in
            data.copyBytes(to: dest, from: offset..<(offset + 2))
        }
        return Int16(bigEndian: raw)
    }
}

File: android/app/src/main/java/com/openburnbar/data/media/MediaPacketCodec.kt
package com.openburnbar.data.media

import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * Length-prefixed binary codec for Mercury media frames. 1:1 port of
 * `MediaPacketCodec.swift`. Same outer layout as `IrohRelayFrameCodec`
 * so an audit reader can locate frame boundaries without knowing
 * whether the payload is JSON chat or a binary media frame.
 *
 * ```
 * [ u32 BE total payload length ][ MediaFrame.HEADER_BYTE_COUNT header ][ optional cursor ][ payload ]
 * ```
 */
class MediaPacketCodec(
    val maxPayloadBytes: Int = DEFAULT_MAX_PAYLOAD_BYTES,
) {
    sealed class CodecError(message: String) : RuntimeException(message) {
        object EnvelopeTooShort : CodecError("envelope too short")
        data class PayloadTooLarge(val actual: Int, val max: Int) :
            CodecError("payload too large: $actual > $max")
        object HeaderTruncated : CodecError("header truncated")
        data class UnknownKind(val raw: Byte) : CodecError("unknown frame kind: $raw")
        object CursorTruncated : CodecError("cursor metadata truncated")
    }

    data class Decoded(val frame: MediaFrame, val consumed: Int)

    fun encode(frame: MediaFrame): ByteArray {
        val cursorByteCount = if (frame.flags.contains(MediaFrame.Flags.HAS_CURSOR_METADATA)) {
            MediaFrame.CURSOR_METADATA_BYTE_COUNT
        } else {
            0
        }
        val totalPayloadCount = MediaFrame.HEADER_BYTE_COUNT + cursorByteCount + frame.payload.size
        if (totalPayloadCount > maxPayloadBytes) {
            throw CodecError.PayloadTooLarge(totalPayloadCount, maxPayloadBytes)
        }
        val buffer = ByteBuffer.allocate(4 + totalPayloadCount).order(ByteOrder.BIG_ENDIAN)
        buffer.putInt(totalPayloadCount)
        buffer.put(frame.kind.rawValue)
        buffer.put(frame.flags.rawValue)
        buffer.putInt(frame.gopID.toInt())
        buffer.putInt(frame.frameIndex.toInt())
        buffer.putLong(frame.presentationTimestampMillis.toLong())
        if (frame.flags.contains(MediaFrame.Flags.HAS_CURSOR_METADATA)) {
            val cursor = frame.cursor ?: MediaFrame.CursorMetadata(0, 0)
            buffer.putShort(cursor.x)
            buffer.putShort(cursor.y)
        }
        buffer.put(frame.payload)
        return buffer.array()
    }

    fun decode(envelope: ByteArray): Decoded {
        val lengthPrefixBytes = 4
        if (envelope.size < lengthPrefixBytes + MediaFrame.HEADER_BYTE_COUNT) {
            throw CodecError.EnvelopeTooShort
        }
        val buffer = ByteBuffer.wrap(envelope).order(ByteOrder.BIG_ENDIAN)
        val totalPayloadCount = buffer.int
        if (totalPayloadCount > maxPayloadBytes) {
            throw CodecError.PayloadTooLarge(totalPayloadCount, maxPayloadBytes)
        }
        val totalEnvelopeBytes = lengthPrefixBytes + totalPayloadCount
        if (envelope.size < totalEnvelopeBytes) {
            throw CodecError.HeaderTruncated
        }
        val kindByte = buffer.get()
        val kind = MediaFrame.Kind.fromRaw(kindByte) ?: throw CodecError.UnknownKind(kindByte)
        val flagsByte = buffer.get()
        val gopID = buffer.int.toUInt()
        val frameIndex = buffer.int.toUInt()
        val pts = buffer.long.toULong()

        var payloadStart = lengthPrefixBytes + MediaFrame.HEADER_BYTE_COUNT
        val cursor = if (MediaFrame.Flags(flagsByte).contains(MediaFrame.Flags.HAS_CURSOR_METADATA)) {
            val cursorEnd = payloadStart + MediaFrame.CURSOR_METADATA_BYTE_COUNT
            if (cursorEnd > lengthPrefixBytes + totalPayloadCount) {
                throw CodecError.CursorTruncated
            }
            val x = buffer.short
            val y = buffer.short
            payloadStart = cursorEnd
            MediaFrame.CursorMetadata(x, y)
        } else {
            null
        }
        val payloadEnd = lengthPrefixBytes + totalPayloadCount
        val payload = envelope.copyOfRange(payloadStart, payloadEnd)

        return Decoded(
            frame = MediaFrame(
                kind = kind,
                flags = MediaFrame.Flags(flagsByte),
                gopID = gopID,
                frameIndex = frameIndex,
                presentationTimestampMillis = pts,
                cursor = cursor,
                payload = payload,
            ),
            consumed = totalEnvelopeBytes,
        )
    }

    companion object {
        /** Hard ceiling on a single media frame. Matches iroh-blobs default chunk size. */
        const val DEFAULT_MAX_PAYLOAD_BYTES: Int = 256 * 1024
    }
}
